File: src/ChpokkWeb/SystemFiles/Templates/ProjectTemplates/CSharp/Silverlight/2052/BusinessApplication/BA.Web/Services/UserRegistrationService.cs
namespace $safeprojectname$
{
    using System;
    using System.Collections.Generic;
    using System.ComponentModel.DataAnnotations;
    using System.ServiceModel.DomainServices.Hosting;
    using System.ServiceModel.DomainServices.Server;
    using System.Web.Profile;
    using System.Web.Security;
    using $safeprojectname$.Resources;

    // TODO: 在部署应用程序时切换为安全终结点。
    //       只应使用 https 传递用户名和密码。
    //       为此，请将 EnableClientAccessAttribute 的 RequiresSecureEndpoint 属性设置为 true。
    //   
    //       [EnableClientAccess(RequiresSecureEndpoint = true)]
    //
    //       有关对域服务使用 https 的更多信息可在 MSDN 上找到。

    /// <summary>
    /// 负责注册用户的域服务。
    /// </summary>
    [EnableClientAccess]
    public class UserRegistrationService : DomainService
    {
        /// <summary>
        /// 默认情况下要将用户添加到的角色。
        /// </summary>
        public const string DefaultRole = "Registered Users";

        //// 注意: 这是可用于启动应用程序的示例代码。
        //// 在成品代码中，您应提供 CAPTCHA 控制功能或验证用户的电子邮件地址，从而针对拒绝服务攻击提供缓解措施。

        /// <summary>
        /// 使用提供的 <see cref="RegistrationData"/> 和 <paramref name="password"/> 添加新用户。
        /// </summary>
        /// <param name="user">此用户的注册信息。</param>
        /// <param name="password">新用户的密码。</param>
        [Invoke(HasSideEffects = true)]
        public CreateUserStatus CreateUser(RegistrationData user,
            [Required(ErrorMessageResourceName = "ValidationErrorRequiredField", ErrorMessageResourceType = typeof(ValidationErrorResources))]
            [RegularExpression("^.*[^a-zA-Z0-9].*$", ErrorMessageResourceName = "ValidationErrorBadPasswordStrength", ErrorMessageResourceType = typeof(ValidationErrorResources))]
            [StringLength(50, MinimumLength = 7, ErrorMessageResourceName = "ValidationErrorBadPasswordLength", ErrorMessageResourceType = typeof(ValidationErrorResources))]
            string password)
        {
            if (user == null)
            {
                throw new ArgumentNullException("user");
            }

            // 请在创建用户之前运行，以确保角色启用且默认角色可用。
            //
            // 如果角色管理器存在问题，则现在失败要好于在创建用户之后失败。
            if (!Roles.RoleExists(UserRegistrationService.DefaultRole))
            {
                Roles.CreateRole(UserRegistrationService.DefaultRole);
            }

            // 注意: ASP.NET 默认情况下使用 SQL Server Express 创建用户数据库。
            // 如果您未安装 SQL Server Express，则 CreateUser 会失败。
            MembershipCreateStatus createStatus;
            Membership.CreateUser(user.UserName, password, user.Email, user.Question, user.Answer, true, null, out createStatus);

            if (createStatus != MembershipCreateStatus.Success)
            {
                return UserRegistrationService.ConvertStatus(createStatus);
            }

            // 将该用户分配给默认角色。
            // 如果禁用角色管理，则此操作将失败。
            Roles.AddUserToRole(user.UserName, UserRegistrationService.DefaultRole);

            // 设置友好名称(配置文件设置)。
            // 如果 web.config 配置不正确，则此操作将失败。
            ProfileBase profile = ProfileBase.Create(user.UserName, true);
            profile.SetPropertyValue("FriendlyName", user.FriendlyName);
            profile.Save();

            return CreateUserStatus.Success;
        }

        private static CreateUserStatus ConvertStatus(MembershipCreateStatus createStatus)
        {
            switch (createStatus)
            {
                case MembershipCreateStatus.Success: return CreateUserStatus.Success;
                case MembershipCreateStatus.InvalidUserName: return CreateUserStatus.InvalidUserName;
                case MembershipCreateStatus.InvalidPassword: return CreateUserStatus.InvalidPassword;
                case MembershipCreateStatus.InvalidQuestion: return CreateUserStatus.InvalidQuestion;
                case MembershipCreateStatus.InvalidAnswer: return CreateUserStatus.InvalidAnswer;
                case MembershipCreateStatus.InvalidEmail: return CreateUserStatus.InvalidEmail;
                case MembershipCreateStatus.DuplicateUserName: return CreateUserStatus.DuplicateUserName;
                case MembershipCreateStatus.DuplicateEmail: return CreateUserStatus.DuplicateEmail;
                default: return CreateUserStatus.Failure;
            }
        }
    }

    /// <summary>
    /// 可以从 <see cref="UserRegistrationService.CreateUser"/> 返回的值的枚举
    /// </summary>
    public enum CreateUserStatus
    {
        Success = 0,
        InvalidUserName = 1,
        InvalidPassword = 2,
        InvalidQuestion = 3,
        InvalidAnswer = 4,
        InvalidEmail = 5,
        DuplicateUserName = 6,
        DuplicateEmail = 7,
        Failure = 8,
    }
}

File: src/ChpokkWeb/SystemFiles/Templates/ProjectTemplates/VisualBasic/Silverlight/3082/BusinessApplication/BA.Web/Services/UserRegistrationService.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel.DataAnnotations
Imports System.ServiceModel.DomainServices.Hosting
Imports System.ServiceModel.DomainServices.Server
Imports System.Web.Profile
Imports System.Web.Security
Namespace Web

    ' TODO: cambie a un extremo seguro al implementar la aplicación.
    '       El nombre de usuario y la contraseña solo se deben pasar mediante https.
    '       Para ello, establezca la propiedad RequiresSecureEndpoint de EnableClientAccessAttribute en true.
    '
    '       <EnableClientAccess(RequiresSecureEndpoint:=True)>
    '
    '       En MSDN encontrará más información sobre el uso de https con un servicio de dominio.

    ''' <summary>
    ''' Servicio de dominio responsable del registro de los usuarios.
    ''' </summary>
    <EnableClientAccess()> _
    Public Class UserRegistrationService
        Inherits DomainService

        ''' <summary>
        ''' Rol al que se agregará a los usuarios de forma predeterminada.
        ''' </summary>
        Public Const DefaultRole As String = "Registered Users"

        ' NOTA: éste es un código de ejemplo para poner en marcha la aplicación.
        ' En el código de producción debería proporcionar una estrategia de reducción frente a un ataque de denegación de servicio al proporcionar funcionalidad de control CAPTCHA o comprobar la dirección de correo electrónico del usuario.

        ''' <summary>
        ''' Agrega un nuevo usuario con los <see cref="RegistrationData"/> y la <paramref name="password"/> proporcionados.
        ''' </summary>
        ''' <param name="user">Información de registro de este usuario.</param>
        ''' <param name="password">Contraseña del nuevo usuario.</param>
        <System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Performance", "CA1822:MarkMembersAsStatic")> _
        <Invoke(HasSideEffects:=True)> _
        Public Function CreateUser(ByVal user As RegistrationData,
            <Required(ErrorMessageResourceName:="ValidationErrorRequiredField", ErrorMessageResourceType:=GetType(ValidationErrorResources))> _
            <RegularExpression("^.*[^a-zA-Z0-9].*$", ErrorMessageResourceName:="ValidationErrorBadPasswordStrength", ErrorMessageResourceType:=GetType(ValidationErrorResources))> _
            <StringLength(50, MinimumLength:=7, ErrorMessageResourceName:="ValidationErrorBadPasswordLength", ErrorMessageResourceType:=GetType(ValidationErrorResources))> _
            ByVal password As String) As CreateUserStatus
            If user Is Nothing Then
                Throw New ArgumentNullException("user")
            End If

            ' Ejecute esto ANTES de crear el usuario para asegurarse de que los roles están habilitados y el rol predeterminado está disponible.
            '
            ' Si hay un problema con el administrador de roles, es mejor que se produzca un error ahora y no una vez creado el usuario.
            If Not Roles.RoleExists(UserRegistrationService.DefaultRole) Then
                Roles.CreateRole(UserRegistrationService.DefaultRole)
            End If

            ' NOTA: ASP.NET utiliza SQL Server Express de forma predeterminada para crear la base de datos de usuarios. 
            ' Si no tiene instalado SQL Server Express, se producirá un error de CreateUser.
            Dim createStatus As MembershipCreateStatus
            Membership.CreateUser(user.UserName, password, user.Email, user.Question, user.Answer, True, Nothing, createStatus)

            If createStatus <> MembershipCreateStatus.Success Then
                Return UserRegistrationService.ConvertStatus(createStatus)
            End If

            ' Asignar al rol predeterminado
            ' Se *puede* producir un error, pero solo si la administración de roles está deshabilitada
            Roles.AddUserToRole(user.UserName, UserRegistrationService.DefaultRole)

            ' Establecer su nombre descriptivo (establecimiento de perfil)
            ' Se *puede* producir un error, pero solo si web.config está configurado incorrectamente 
            Dim profile As ProfileBase = ProfileBase.Create(user.UserName, True)
            profile.SetPropertyValue("FriendlyName", user.FriendlyName)
            profile.Save()

            Return CreateUserStatus.Success
        End Function

        Private Shared Function ConvertStatus(ByVal createStatus As MembershipCreateStatus) As CreateUserStatus
            Select Case createStatus
                Case MembershipCreateStatus.Success
                    Return CreateUserStatus.Success
                Case MembershipCreateStatus.InvalidPassword
                    Return CreateUserStatus.InvalidPassword
                Case MembershipCreateStatus.InvalidEmail
                    Return CreateUserStatus.InvalidEmail
                Case MembershipCreateStatus.InvalidAnswer
                    Return CreateUserStatus.InvalidAnswer
                Case MembershipCreateStatus.InvalidQuestion
                    Return CreateUserStatus.InvalidQuestion
                Case MembershipCreateStatus.InvalidUserName
                    Return CreateUserStatus.InvalidUserName
                Case MembershipCreateStatus.DuplicateUserName
                    Return CreateUserStatus.DuplicateUserName
                Case MembershipCreateStatus.DuplicateEmail
                    Return CreateUserStatus.DuplicateEmail
                Case Else
                    Return CreateUserStatus.Failure
            End Select
        End Function
    End Class

    ''' <summary>
    ''' Enumeración de los valores que se pueden devolver desde <see cref="UserRegistrationService.CreateUser"/>
    ''' </summary>
    Public Enum CreateUserStatus
        Success = 0
        InvalidUserName = 1
        InvalidPassword = 2
        InvalidQuestion = 3
        InvalidAnswer = 4
        InvalidEmail = 5
        DuplicateUserName = 6
        DuplicateEmail = 7
        Failure = 8
    End Enum
End Namespace